### PR TITLE
Fixes to codebase for MSVC 2017 

### LINKED ISSUE #20598: Preparations for MSVC 2017
Update codebase (remove any issues) for compatibility with MSVC 2017.

### PR DESCRIPTION
**Pull request by LamarMoore**
Modifications to the codebase to facilitate building with MSVC 2017

**To test:**

- build with MSVC 2017 and ensure everything passes
- run all unit tests
- test output package created with CPACK

Fixes #20598. 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/mantid/api/src/Exports/Axis.cpp [hunk 1]
```diff
@@ -24,7 +24,7 @@ using Mantid::Kernel::Unit_sptr;
 using Mantid::specnum_t;
 using namespace boost::python;
 
-GET_POINTER_SPECIALIZATION(Axis)
+// GET_POINTER_SPECIALIZATION(Axis)
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/api/src/Exports/Axis.cpp]: I think the removal of these is the cause of the MSVC build errors. As `GET_POINTER_SPECIALIZATION` is our own macro can we just make this a no op for 2017?

File: Framework/Algorithms/src/DiffractionFocussing2.cpp [hunk 2]
```diff
@@ -307,9 +307,11 @@ void DiffractionFocussing2::exec() {
                    std::divides<double>());
     // Now multiply by the number of spectra in the group
     std::transform(Yout.begin(), Yout.end(), Yout.begin(),
-                   std::bind2nd(std::multiplies<double>(), groupSize));
+                   std::bind2nd(std::multiplies<double>(),
+                                static_cast<double>(groupSize)));
```
Comment (quantumsteve) [inline Framework/Algorithms/src/DiffractionFocussing2.cpp]: Would you replace the deprecated [`std::bind2nd`](http://en.cppreference.com/w/cpp/utility/functional/bind12) with either `std::bind` or a lambda expression?

File: Framework/PythonInterface/mantid/api/src/Exports/Axis.cpp (hunk 1 above)
Comment (LamarMoore) [reply Framework/PythonInterface/mantid/api/src/Exports/Axis.cpp]: I guess so but it was not disabled in all cases. But I will see what happens when I do what you suggest.

**Comment by martyngigg** [comment]
Just for clarity in the conversation. We have found that the current build gives a warning "Unknown compiler version" due to this line in the boost compiler config: https://github.com/boostorg/config/blob/boost-1.59.0/include/boost/config/compiler/visualc.hpp#L291.

We'll upgrade boost and patch this warning away. This has already been done in the boost [develop](https://github.com/boostorg/config/commit/a0147b73e8bde8f4b2c19c96166353dba8fa9914) branch.

**Comment by martyngigg** [comment]
I've also found a tiny issue with the `command-prompt.bat` file generated in the `build` directory - it still links to VS2015.

**Review by martyngigg — CHANGES_REQUESTED** [review]
I've also found a tiny issue with the command-prompt.bat file generated in the build directory - it still links to VS2015.

File: buildconfig/CMake/MSVCSetup.cmake [hunk 3]
```diff
@@ -89,6 +89,11 @@ set ( CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/bin )
 set ( WINDOWS_BUILDCONFIG ${PROJECT_SOURCE_DIR}/buildconfig/windows )
 configure_file ( ${WINDOWS_BUILDCONFIG}/buildenv.bat.in ${PROJECT_BINARY_DIR}/buildenv.bat @ONLY )
 configure_file ( ${WINDOWS_BUILDCONFIG}/command-prompt.bat ${PROJECT_BINARY_DIR}/command-prompt.bat @ONLY )
+
+get_filename_component ( MSVC_IDE_LOCATION "${CMAKE_CXX_COMPILER}" DIRECTORY )
```
Comment (martyngigg) [inline buildconfig/CMake/MSVCSetup.cmake]: Can we make this dependent on the generator? This no longer works in VS2015. I'd like to keep 2015 running for a bit that we aren't required to switch all of the buildservers immediately on the same day.

File: buildconfig/CMake/MSVCSetup.cmake [hunk 4]
```diff
@@ -88,12 +88,18 @@ set ( CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/bin )
 ###########################################################################
 set ( WINDOWS_BUILDCONFIG ${PROJECT_SOURCE_DIR}/buildconfig/windows )
 configure_file ( ${WINDOWS_BUILDCONFIG}/buildenv.bat.in ${PROJECT_BINARY_DIR}/buildenv.bat @ONLY )
-configure_file ( ${WINDOWS_BUILDCONFIG}/command-prompt.bat ${PROJECT_BINARY_DIR}/command-prompt.bat @ONLY )
 
-get_filename_component ( MSVC_IDE_LOCATION "${CMAKE_CXX_COMPILER}" DIRECTORY )
-get_filename_component ( MSVC_IDE_LOCATION "${MSVC_IDE_LOCATION}/../../../../../../.." ABSOLUTE )
-set ( MSVC_IDE_LOCATION "${MSVC_IDE_LOCATION}/Common7/IDE")
+if ( MSVC_VERSION EQUAL "1911" )
```
Comment (martyngigg) [inline buildconfig/CMake/MSVCSetup.cmake]: Should we switch this around and check `< 1911` and do the VS140 stuff else do the VS2017 and beyond steps? In the future I presume newer versions will follow the VS2017 convention and then we won't have to update this code.

Comment (LamarMoore) [reply buildconfig/CMake/MSVCSetup.cmake]: Good point

**Review by martyngigg — APPROVED** [review]

**Pull request merged**